packages/tangle/approvalweightmanager.go

```diff
@@ -86,6 +87,70 @@ func (a *ApprovalWeightManager) weightsPerEpoch(branchID ledgerstate.BranchID) *
 	return weightsPerEpoch
 }
 
+func (a *ApprovalWeightManager) onMarkerConfirmed(marker *markers.Marker, messageID MessageID) {
+	// walk through the marker's pastcone to update FinalizedApprovalWeight
+	finalizedWalker := walker.New()
```

Comment:
Instead of doing all this manually you can simply use `WalkMessageAndMetadata`. This should also simplify `propagateFinalizedApprovalWeight` as you don't need to consume.